app/src/main/java/com/beemdevelopment/aegis/AssignIconsActivity.java

```diff
@@ -0,0 +1,247 @@
+import android.view.View;
+import android.widget.Toast;
+
+import androidx.activity.OnBackPressedCallback;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.recyclerview.widget.LinearLayoutManager;
+import androidx.recyclerview.widget.RecyclerView;
+
+import com.beemdevelopment.aegis.helpers.MetricsHelper;
+import com.beemdevelopment.aegis.icons.IconPack;
+import com.beemdevelopment.aegis.ui.AegisActivity;
+import com.beemdevelopment.aegis.ui.dialogs.Dialogs;
+import com.beemdevelopment.aegis.ui.dialogs.IconPickerDialog;
+import com.beemdevelopment.aegis.ui.models.AssignIconEntry;
+import com.beemdevelopment.aegis.ui.views.AssignIconAdapter;
+import com.beemdevelopment.aegis.ui.views.IconAdapter;
+import com.beemdevelopment.aegis.util.IOUtils;
+import com.beemdevelopment.aegis.vault.VaultEntry;
+import com.bumptech.glide.Glide;
+import com.bumptech.glide.ListPreloader;
+import com.bumptech.glide.RequestBuilder;
+import com.bumptech.glide.integration.recyclerview.RecyclerViewPreloader;
+import com.bumptech.glide.load.engine.DiskCacheStrategy;
+import com.bumptech.glide.util.ViewPreloadSizeProvider;
+import com.google.android.material.bottomsheet.BottomSheetDialog;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+
+public class AssignIconsActivity extends AegisActivity implements AssignIconAdapter.Listener {
+    private AssignIconAdapter _adapter;
+    private ArrayList<AssignIconEntry> _entries = new ArrayList<>();
+    private RecyclerView _entriesView;
+    private AssignIconsActivity.BackPressHandler _backPressHandler;
+    private ViewPreloadSizeProvider<AssignIconEntry> _preloadSizeProvider;
+    private IconPack _favoriteIconPack;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        if (abortIfOrphan(savedInstanceState)) {
+            return;
+        }
+
+        setContentView(R.layout.activity_assign_icons);
+        setSupportActionBar(findViewById(R.id.toolbar));
+        if (getSupportActionBar() != null) {
+            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
+            getSupportActionBar().setDisplayShowHomeEnabled(true);
+        }
+
+        ArrayList<UUID> assignIconEntriesIds = (ArrayList<UUID>) getIntent().getSerializableExtra("entries");
+        for (UUID entryId: assignIconEntriesIds) {
+            VaultEntry vaultEntry = _vaultManager.getVault().getEntryByUUID(entryId);
+            _entries.add(new AssignIconEntry(vaultEntry));
+        }
+
+        _backPressHandler = new AssignIconsActivity.BackPressHandler();
+        getOnBackPressedDispatcher().addCallback(this, _backPressHandler);
+
+        AssignIconsActivity.IconPreloadProvider modelProvider = new AssignIconsActivity.IconPreloadProvider();
```

Comment:
This is still not quite right. It appears that the preloader is not used at all and the culprit seems to be the ``NestedScrollView`` that wraps the ``RecyclerView``. We can probably just get rid of that ``NestedScrollView``, as it breaks the rest of the functionality of RecyclerView as well. I see that we're doing this in a few other places throughout the codebase, we should probably review those cases some other time.

I also think we need two preloaders for Glide. One for the existing icons and one for the new icons.

Here are some of the (WIP) changes I made while testing, in case it's useful: https://github.com/alexbakker/Aegis/commit/56854b2178e1b1cef22e06f374d324c78d160edd